ESQL: Add ability to perform date math

bpintea opened the pull request:
This adds the ability to perform date math: allow date types be updated with a time span (a duration or a period).

Ex. : `row n = now() | eval then = n - 1 year + 2 minutes`

Closes #98402.

bpintea committed:
Add ability to perform date math

This adds the ability to perform date math: allow date types be updated
with a time span (a duration or a period).

Ex. : `row n = now() | eval then = n - 1 year + 2 minutes`

elasticsearchmachine commented:
Hi @bpintea, I've created a changelog YAML for you.

Commit:
Update docs/changelog/98870.yaml

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -0,0 +1,81 @@
+        DataType leftType = left().dataType();
+        DataType rightType = right().dataType();
+        if (EsqlDataTypes.isDateTime(leftType) || EsqlDataTypes.isDateTime(rightType)) {
+            Expression dateTime = argumentOfType(dt -> dt == DataTypes.DATETIME);
```
nit: intention of the if-statement's body is not immediately clear IMO; maybe a comment like `determine whether there's  one datetime argument and one time period/duration argument`

bpintea replied:
Added a comment.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -0,0 +1,81 @@
+
+    protected static final ZoneId DEFAULT_TZ = DateUtils.UTC;
+
+    private final DatetimeArithmeticEvaluator datetimes;
```
nit: naming somewhat unclear IMO, maybe renaming to e.g. `datetimesEvaluator` would be clearer?

bpintea replied:
I wanted to stay consistent with the name of the other type evaluators (both fields - like `ints` and evaluator class name - like `SubDatetimesEvaluator`).

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -0,0 +1,81 @@
+        if (EsqlDataTypes.isDateTime(leftType) || EsqlDataTypes.isDateTime(rightType)) {
+            Expression dateTime = argumentOfType(dt -> dt == DataTypes.DATETIME);
+            Expression nonDateTime = argumentOfType(dt -> dt != DataTypes.DATETIME);
+            if (dateTime == null || nonDateTime == null || EsqlDataTypes.isTemporalAmount(nonDateTime.dataType()) == false) {
```
question: instead of using `nonDateTime` together with `isTemporalAmount`, wouldn't it be easier to just use `argumentOfType(EsqlDataTypes::isTemporalAmount`, like below in line 73?

bpintea replied:
Updated

alex-spies reviewed (approved):
LGTM, I only have minor remarks which are all meant to be optional.

Additionally, maybe it would make sense to add the following tests:
- unit tests tests for `period + datetime` and `duration + datetime`
- unit tests for `period - datetime` and `duration - datetime`
- unit or csv tests for weird edge cases, like date math for leap years, adding days or hours past the end of a month, etc. Although this may be less important since we know that we can trust the date math functions from the `ql` package.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/Add.java:
```diff
@@ -75,4 +82,9 @@ public static long processUnsignedLongs(long lhs, long rhs) {
+
+    @Evaluator(extraName = "Datetimes", warnExceptions = { ArithmeticException.class, DateTimeException.class })
+    static long processDatetimes(long datetime, @Fixed TemporalAmount temporalAmount) {
+        return asMillis(asDateTime(datetime, DEFAULT_TZ).plus(temporalAmount));
```
Why use UTC and not the request timezone?

bpintea replied:
I've added a comment. 
Both timestamps that come from ES as well as those converted by our functions are in UTC.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java:
```diff
@@ -52,6 +52,14 @@ public static ZonedDateTime asDateTime(long millis) {
         return ZonedDateTime.ofInstant(Instant.ofEpochMilli(millis), UTC);
     }
 
+    public static ZonedDateTime asDateTime(long millis, ZoneId zoneId) {
```
Since you always pass UTC, why not use the above?

bpintea replied:
Remanent from a first attempt to use session's zoneID. Updated to use the existing one.

costin reviewed (commented):
The main comment I have is around the use of timezones.
Date functions need to be time zone aware yet here the UTC tz is used - that makes sense only if the dates are considered to be coming from ES and thus they originate in UTC but the output is converted into the user TZ.
If that is not the case, this should be clarified.

Lastly some docs help - I would reuse the structure and content from SQL since its fairly similar.

bpintea committed:
Add unit testing. Make temporal units negatable

Add Add/Sub unit tests
Make temporal units negatable.
Rename data type tests methods.
More tests.

bpintea committed:
Merge branch 'main' into esql/feature/date_math

bpintea commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/date.csv-spec:
```diff
@@ -453,3 +453,128 @@ emp_no:integer  |  birth_date:datetime       | birth_month:keyword
+2100-01-01T01:01:01.000Z |2104-04-16T01:01:01.000Z
+;
+
+datePlusDuration
```
Good point. I've extended the PR to introduce support for negating temporal values.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java:
```diff
@@ -137,6 +137,14 @@ public static boolean isPrimitive(DataType t) {
         return t != OBJECT && t != NESTED;
     }
 
+    public static boolean isDateTime(DataType t) {
```
I've updated the method name (to `isDateTimeOrTemporal()`).

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -0,0 +1,81 @@
+        DataType rightType = right().dataType();
+        if (EsqlDataTypes.isDateTime(leftType) || EsqlDataTypes.isDateTime(rightType)) {
+            Expression dateTime = argumentOfType(dt -> dt == DataTypes.DATETIME);
+            Expression nonDateTime = argumentOfType(dt -> dt != DataTypes.DATETIME);
```
> looking at the evaluators, this does not seem to be supported.

It should be. I've added some more tests.

bpintea commented:
> The main comment I have is around the use of timezones.
> If that is not the case, this should be clarified.

I've added a comment.

> Lastly some docs help - I would reuse the structure and content from SQL since its fairly similar.

Will follow with a subsequent PR.

bpintea commented:
> unit tests tests for period + datetime and duration + datetime

added.

> unit tests for period - datetime and duration - datetime

Good point, thanks. Increased the Sub type resolution checks to prevent this.

> unit or csv tests for weird edge cases, like date math for leap years, adding days or hours past the end of a month, etc. Although this may be less important since we know that we can trust the date math functions from the ql package.

Indeed, left this out.

costin reviewed (approved):
LGTM

alex-spies reviewed (approved):
Still LGTM, only added very minor remarks (all supposed to be completely optional). Did not look too closely into what appeared to be refactors of the test that probably became necessary due to https://github.com/elastic/elasticsearch/pull/98890.

bpintea committed:
Address review comments

Address review comments

bpintea committed:
Merge branch 'main' into esql/feature/date_math

bpintea commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/SubTests.java:
```diff
@@ -71,9 +81,32 @@ public static Iterable<Object[]> parameters() {
+                List.of(new TypedData(lhs, DataTypes.DATETIME, "lhs"), new TypedData(rhs, EsqlDataTypes.DATE_PERIOD, "rhs")),
+                "SubDatetimesEvaluator[lhs=Attribute[channel=0], rhs=Attribute[channel=1]]",
+                DataTypes.DATETIME,
+                equalTo(asMillis(asDateTime(lhs).minus(rhs)))
```
We would want that, but the `TestCaseSupplier`s are built statically and the `expectedValue()` is inherited/overridden; otherwise we could have used the `resultMatcher()` method directly (which would have been even nicer).

bpintea commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/NegTests.java:
```diff
@@ -112,14 +137,47 @@ public void testEdgeCases() {
+            Neg neg = new Neg(Source.EMPTY, field("val", typeOf(val)));
+            return toJavaObject(evaluator(neg).get().eval(row(List.of(val))), 0);
+        } else {
+            Neg neg = new Neg(Source.EMPTY, new Literal(Source.EMPTY, val, typeOf(val)));
```
This helper function would still need to be invoked from `process()`, since I'd like to avoid having the test cases "decide" how to "process" the negation. This being a test, I've just added a small comment on the else branch.
